### PR TITLE
[WIP] type 2 finufft consolidated

### PR DESCRIPTION
**Pull request by eickenberg**
Took a while to figure out but the most natural setting for isign in type 2 is -1, which is also the default in finufft.

TODO:
- [x] add forward tests for 1D and 3D
- [x] implement backward for values (working for 2D)
- [x] implement backward for positions (working for 2D)
- [x] tests for gradients in 2D
- [x] tests for other dimensionalities

### CODE REVIEW

File: pytorch_finufft/functional.py [hunk 1]
```diff
@@ -836,3 +836,205 @@ def backward(  # type: ignore[override]
...
+        # slices = tuple(slice(start, end) for start, end in zip(start_points, end_points))
+
+        # # CPU idiosyncracy that needs to be done differently
+        # k_ramps = torch.from_numpy(np.mgrid[slices], dtype=points.dtype)
+
+        grad_points = grad_targets = None
+
+        ## From type 1, commenting for now to understand whether needed
+        # if any(ctx.needs_input_grad) and _mode_ordering:
+        #     grad_output = torch.fft.fftshift(grad_output)
```
Comment (WardBrian) [inline pytorch_finufft/functional.py:986]: Is it surprising that this seems to be not needed?

Comment (eickenberg) [reply pytorch_finufft/functional.py:986]: this is a good point. It can't be applied like here because type 2 output is on unstructured points. However, it needs to go somewhere. 

Your comment made me discover that we are not testing different mode ordering in the backwards passes

Comment (eickenberg) [reply pytorch_finufft/functional.py:986]: wait no we are ...

Comment (WardBrian) [reply pytorch_finufft/functional.py:986]: Are we not? I thought one of the things we are combinatorially testing is fftshift true/false

File: pytorch_finufft/functional.py [hunk 2]
```diff
@@ -836,3 +836,201 @@ def backward(  # type: ignore[override]
...
+        targets: torch.Tensor,
+        out: Optional[torch.Tensor] = None,
+        finufftkwargs: Dict[str, Union[int, float]] = {},
+    ) -> torch.Tensor:
+        """
+        Evaluates the Type 2 NUFFT on the inputs.
+
+        NOTE: By default, the ordering is set to match that of Pytorch,
+         Numpy, and Scipy's FFT APIs. To match the mode ordering
+         native to FINUFFT, set fftshift=True.
```
Comment (WardBrian) [inline pytorch_finufft/functional.py:860]: Docstring mentions fftshift which is gone now

File: pytorch_finufft/functional.py [hunk 3]
```diff
@@ -836,3 +836,201 @@ def backward(  # type: ignore[override]
...
+        ```
+
+        Parameters
+        ----------
+        ctx : Any
+            Pytorch context objecy
+        points_x : torch.Tensor
+            The non-uniform points x_j
+        points_y : torch.Tensor
+            The non-uniform points y_j
```
Comment (WardBrian) [inline pytorch_finufft/functional.py:877]: Out of date

File: pytorch_finufft/functional.py [hunk 4]
```diff
@@ -836,3 +836,201 @@ def backward(  # type: ignore[override]
...
+        targets : torch.Tensor
+            The target Fourier mode coefficients f[k1, k2]
+        out : Optional[torch.Tensor], optional
+            Array to take the result in-place, by default None
+        fftshift : bool
+            If True centers the 0 mode in the resultant torch.Tensor, by default False
```
Comment (WardBrian) [inline pytorch_finufft/functional.py:883]: Out of date

File: pytorch_finufft/functional.py [hunk 5]
```diff
@@ -836,3 +836,201 @@ def backward(  # type: ignore[override]
...
+
+        Returns
+        -------
+        torch.Tensor
+            The resultant array c[j]
+
+        Raises
+        ------
+        ValueError
+            In the case of conflicting specification of the wave-mode ordering.
```
Comment (WardBrian) [inline pytorch_finufft/functional.py:897]: Out of date

File: pytorch_finufft/functional.py [hunk 6]
```diff
@@ -836,3 +836,201 @@ def backward(  # type: ignore[override]
...
+        _mode_ordering = finufftkwargs.pop("modeord", 1)
+        _i_sign = finufftkwargs.pop("isign", -1)
+
+        # if fftshift:
+        #     if _mode_ordering != 1:  # This seems like it is the wrong way round???????
+        #         raise ValueError(
+        #             "Double specification of ordering; only one of fftshift and "
+        #             "modeord should be provided."
+        #         )
+        #     _mode_ordering = 0
```
Comment (WardBrian) [inline pytorch_finufft/functional.py:916]:
```suggestion
```

File: pytorch_finufft/functional.py [hunk 7]
```diff
@@ -836,3 +836,201 @@ def backward(  # type: ignore[override]
...
+
+        ndim = points.shape[0]
+        if _mode_ordering == 1:
+            targets = torch.fft.fftshift(targets, dim=tuple(range(-ndim, 0)))
```
Comment (WardBrian) [inline pytorch_finufft/functional.py:920]: The `dim` argument here will be useful if we allow batching but for now I wonder if it's necessary. I'm happy either way

File: pytorch_finufft/functional.py [hunk 8]
```diff
@@ -836,3 +836,201 @@ def backward(  # type: ignore[override]
...
+
+        if out is not None:
+            print("In-place results are not yet implemented")
+
+        # TODO -- extend checks to 2d
+        checks._type2_checks(points, targets)
+
+        finufftkwargs = {k: v for k, v in finufftkwargs.items()}
+        _mode_ordering = finufftkwargs.pop("modeord", 1)
+        _i_sign = finufftkwargs.pop("isign", -1)
```
Comment (WardBrian) [inline pytorch_finufft/functional.py:908]: Can you add a comment after each of these lines on whether or not it matches the FINUFFT default?

File: pytorch_finufft/functional.py [hunk 9]
```diff
@@ -836,3 +836,201 @@ def backward(  # type: ignore[override]
...
+        ctx.mode_ordering = _mode_ordering
+        ctx.finufftkwargs = finufftkwargs
+
+        ctx.save_for_backward(points, targets)
+
+        nufft_func = get_nufft_func(ndim, 2, points.device.type)
+
+        finufft_out = nufft_func(
+            *points.data.numpy(),
+            targets.data.numpy(),
```
Comment (WardBrian) [inline pytorch_finufft/functional.py:933]: I believe if you remove the `.data.numpy()` this function is then immediately ready for CUDA

File: pytorch_finufft/functional.py [hunk 10]
```diff
@@ -836,3 +836,201 @@ def backward(  # type: ignore[override]
...
+
+        points, targets = ctx.saved_tensors
+        device = points.device
+
+        # start_points = -(np.array(targets.shape) // 2)
+        # end_points = start_points + targets.shape
+        # slices = tuple(slice(start, end) for start, end in zip(start_points, end_points))
+
+        # # CPU idiosyncracy that needs to be done differently
+        # k_ramps = torch.from_numpy(np.mgrid[slices], dtype=points.dtype)
```
Comment (WardBrian) [inline pytorch_finufft/functional.py:978]:
```suggestion
```

File: pytorch_finufft/functional.py [hunk 11]
```diff
@@ -836,3 +836,201 @@ def backward(  # type: ignore[override]
...
+            end_points = start_points + torch.tensor(targets.shape, device=device)
+            coord_ramps = torch.stack(
+                torch.meshgrid(
+                    *(
+                        torch.arange(start, end, device=device)
+                        for start, end in zip(start_points, end_points)
+                    ),
+                    indexing="ij",
+                )
+            )
```
Comment (WardBrian) [inline pytorch_finufft/functional.py:996]: [optional] if this is exactly the same code as type1, I wonder if it is worth pulling out into a helper function

File: pytorch_finufft/functional.py [hunk 12]
```diff
@@ -836,3 +836,201 @@ def backward(  # type: ignore[override]
...
+
+        if ctx.needs_input_grad[0]:
```
Comment (WardBrian) [inline pytorch_finufft/functional.py:998]: Combine this with the above identical `if` statement

File: pytorch_finufft/functional.py [hunk 13]
```diff
@@ -836,3 +836,201 @@ def backward(  # type: ignore[override]
...
+            ramped_targets = coord_ramps * targets[np.newaxis] * 1j * _i_sign
+            nufft_func = get_nufft_func(ndim, 2, points.device.type)
+
+            grad_points = nufft_func(
+                    *points,
+                    ramped_targets.squeeze(),
+                    isign=_i_sign,
+                    #modeord=_mode_ordering,
```
Comment (WardBrian) [inline pytorch_finufft/functional.py:1006]:
```suggestion
```

File: pytorch_finufft/functional.py [hunk 14]
```diff
@@ -836,3 +836,201 @@ def backward(  # type: ignore[override]
...
+
+        if ctx.needs_input_grad[1]:
+            # wrt. targets
+            nufft_func = get_nufft_func(ndim, 1, points.device.type)
+
+            grad_targets = nufft_func(
+                    *points,
+                    grad_output,
+                    targets.shape,
+                    #modeord=_mode_ordering,
```
Comment (WardBrian) [inline pytorch_finufft/functional.py:1021]:
```suggestion
```

File: pytorch_finufft/functional.py [hunk 15]
```diff
@@ -836,3 +836,201 @@ def backward(  # type: ignore[override]
...
+                    isign=-_i_sign,
+                    **finufftkwargs,
+                )
+            
+            if _mode_ordering == 1:
+                grad_targets = torch.fft.ifftshift(grad_targets, dim=tuple(range(-ndim, 0)))
```
Comment (WardBrian) [inline pytorch_finufft/functional.py:1027]: Similar note to above on the `dim` argument

File: tests/test_1d/test_backward_1d.py [hunk 16]
```diff
@@ -161,3 +161,50 @@ def test_t2_backward_CPU_points(
...
+
+
+@pytest.mark.parametrize("N", Ns)
+@pytest.mark.parametrize("modifier", length_modifiers)
+@pytest.mark.parametrize("fftshift", [False, True])
+@pytest.mark.parametrize("isign", [-1, 1])
+def test_t2_backward_CPU_values(
+    N: int, modifier: int, fftshift: bool, isign: int
+) -> None:
+    check_t2_backward(N, modifier, fftshift, isign, "cpu", False)
```
Comment (WardBrian) [inline tests/test_1d/test_backward_1d.py:200]: If you want, the cuda tests should be obvious extensions here and CI should pick them up for you

**Review by WardBrian — APPROVED** [review]
This looks good overall, thanks for doing the leg work on understanding what's going on.

If there are any comments you agree with but don't want to do yourself, we can merge and I can pick them up in another PR

File: pytorch_finufft/functional.py (hunk 11 above)
Comment (eickenberg) [reply pytorch_finufft/functional.py]: good idea

Comment (eickenberg) [reply pytorch_finufft/functional.py]: done

File: pytorch_finufft/functional.py (hunk 12 above)
Comment (eickenberg) [reply pytorch_finufft/functional.py:963]: removed via implementation of helper function as suggested above

File: pytorch_finufft/functional.py (hunk 13 above)
Comment (eickenberg) [reply pytorch_finufft/functional.py]: done

File: pytorch_finufft/functional.py (hunk 14 above)
Comment (eickenberg) [reply pytorch_finufft/functional.py]: done

File: pytorch_finufft/functional.py (hunk 15 above)
Comment (eickenberg) [reply pytorch_finufft/functional.py]: removed noop stuff

File: tests/test_1d/test_backward_1d.py (hunk 16 above)
Comment (eickenberg) [reply tests/test_1d/test_backward_1d.py:200]: good point, added those

**Review by WardBrian — APPROVED** [review]
Awesome!
